File: python/Facade/movie_element.py
class Amplifier:
    def __init__(self) -> None:
        self.name = "Top-O-Line Amplifier"

    def off(self):
        print(self.name, f"off")

    def on(self):
        print(self.name, "on")


class DvdPlayer:
    def __init__(self) -> None:
        self.name = "Top-O-Line Dvd Player"

    def off(self):
        print(self.name, f"off")

    def on(self):
        print(self.name, "on")

    def play(self, movie: str):
        self.movie = movie
        print(self.name, f'playing "{movie}"')

    def stop(self):
        print(self.name, f'stopped "{self.movie}"')


class Projecter:
    def __init__(self) -> None:
        self.name = "Top-O-Line Projector"

    def off(self):
        print(self.name, f"off")

    def on(self):
        print(self.name, "on")


class TheaterLights:
    def __init__(self) -> None:
        self.name = "Thether Ceiling Lights"

    def dim(self, num: int):
        print(self.name, f"dimming to {num}%")

    def on(self):
        print(self.name, "on")


class Screen:
    def __init__(self) -> None:
        self.name = "Thether Screen"

    def up(self):
        print(self.name, "going up")

    def down(self):
        print(self.name, "going down")


class PopcornPopper:
    def __init__(self) -> None:
        self.name = "Popcron Popper"

    def on(self):
        print(self.name, "on")

    def off(self):
        print(self.name, "off")

    def pop(self):
        print(self.name, "poping popcorn!")

File: c/Facade/movie_element.c
#include "movie_element.h"

void AmplifierNew(Amplifier *self)
{
    strcpy(self->name, "Top-O-Line Amplifier");
    self->Off = AmplifierOff;
    self->On = AmplifierOn;
}
void AmplifierOff(Amplifier *self)
{
    printf("%s off\n", self->name);
}
void AmplifierOn(Amplifier *self)
{
    printf("%s on\n", self->name);
}

void DvdPlayerNew(DvdPlayer *self)
{
    strcpy(self->name, "Top-O-Line Dvd Player");
    self->Off = DvdPlayerOff;
    self->On = DvdPlayerOn;
    self->Play = DvdPlayerPlay;
    self->Stop = DvdPlayerStop;
}
void DvdPlayerOff(DvdPlayer *self)
{
    printf("%s off\n", self->name);
}
void DvdPlayerOn(DvdPlayer *self)
{
    printf("%s on\n", self->name);
}
void DvdPlayerPlay(DvdPlayer *self, const char *movie)
{
    strcpy(self->movie, movie);
    printf("%s playing \"%s\"\n", self->name, self->movie);
}
void DvdPlayerStop(DvdPlayer *self)
{
    printf("%s stopped \"%s\"\n", self->name, self->movie);
}

void ProjectorNew(Projector *self)
{
    strcpy(self->name, "Top-O-Line Projector");
    self->Off = ProjectorOff;
    self->On = ProjectorOn;
}
void ProjectorOff(Projector *self)
{
    printf("%s off\n", self->name);
}
void ProjectorOn(Projector *self)
{
    printf("%s on\n", self->name);
}

void TheaterLightsNew(TheaterLights *self)
{
    strcpy(self->name, "Theater Ceiling Lights");
    self->Dim = TheaterLightsDim;
    self->On = TheaterLightsOn;
}
void TheaterLightsDim(TheaterLights *self, int num)
{
    printf("%s dimming to %d%%\n", self->name, num);
}
void TheaterLightsOn(TheaterLights *self)
{
    printf("%s on\n", self->name);
}

void ScreenNew(Screen *self)
{
    strcpy(self->name, "Theater Screen");
    self->Up = ScreenUp;
    self->Down = ScreenDown;
}
void ScreenUp(Screen *self)
{
    printf("%s going up\n", self->name);
}
void ScreenDown(Screen *self)
{
    printf("%s going down\n", self->name);
}

void PopcornPopperNew(PopcornPopper *self)
{
    strcpy(self->name, "Popcorn Popper");
    self->On = PopcornPopperOn;
    self->Off = PopcornPopperOff;
    self->Pop = PopcornPopperPop;
}
void PopcornPopperOff(PopcornPopper *self)
{
    printf("%s off\n", self->name);
}
void PopcornPopperOn(PopcornPopper *self)
{
    printf("%s on\n", self->name);
}
void PopcornPopperPop(PopcornPopper *self)
{
    printf("%s poping popcorn!\n", self->name);
}